File: resources/ios/SearchFieldRenderer.swift
import SwiftUI

enum SearchFieldRendererEvent: Equatable {
    case change(callbackId: Int, nodeId: Int, value: String)
    case submit(callbackId: Int, nodeId: Int, value: String)

    var wireName: String {
        switch self {
        case .change: "TEXT_CHANGE"
        case .submit: "SUBMIT"
        }
    }
}

struct SearchFieldRendererEvents {
    let send: @MainActor (SearchFieldRendererEvent) -> Void

    static let native = SearchFieldRendererEvents { event in
        switch event {
        case let .change(callbackId, nodeId, value):
            NativeUIBridge.sendTextChangeEvent(callbackId, nodeId: nodeId, text: value)
        case let .submit(callbackId, nodeId, value):
            NativeUIBridge.sendSubmitEvent(callbackId, nodeId: nodeId, text: value)
        }
    }
}

struct SearchFieldRendererConfiguration {
    let nodeID: Int
    let value: String
    let placeholder: String
    let disabled: Bool
    let autocapitalize: String
    let autocorrectPolicy: String
    let syncMode: String
    let debounceMilliseconds: Int
    let accessibilityLabel: String
    let accessibilityHint: String
    let onChangeCallback: Int
    let onSubmitCallback: Int

    init(node: NativeUINode) {
        let props = node.props
        nodeID = node.id
        value = props.getString("value")
        placeholder = props.getString("placeholder")
        disabled = props.getBool("disabled")
        autocapitalize = props.getString("autocapitalize")
        autocorrectPolicy = props.getString("autocorrect_policy", default: "default")
        syncMode = props.getString("sync_mode", default: "live")
        debounceMilliseconds = max(50, props.getInt("debounce_ms", default: 300))
        accessibilityLabel = props.getString("a11y_label")
        accessibilityHint = props.getString("a11y_hint")
        onChangeCallback = props.getCallbackId("on_change")
        onSubmitCallback = props.getCallbackId("on_submit")
    }
}

struct SearchFieldRendererState {
    var configuration: SearchFieldRendererConfiguration
    var draft: String
    var lastCommitted: String
    var pendingServerValue: String?
    var isFocused = false

    init(node: NativeUINode) {
        configuration = SearchFieldRendererConfiguration(node: node)
        draft = configuration.value
        lastCommitted = configuration.value
        pendingServerValue = nil
    }

    mutating func userChanged(_ value: String) -> SearchFieldRendererEvent? {
        guard !configuration.disabled else { return nil }
        draft = value
        return configuration.syncMode == "live" ? commitIfNeeded() : nil
    }

    mutating func focusChanged(_ focused: Bool) -> SearchFieldRendererEvent? {
        isFocused = focused
        return focused ? nil : commitIfNeeded()
    }

    mutating func flush() -> SearchFieldRendererEvent? {
        commitIfNeeded()
    }

    mutating func submit() -> [SearchFieldRendererEvent] {
        guard !configuration.disabled else { return [] }
        var events: [SearchFieldRendererEvent] = []
        if let change = commitIfNeeded() { events.append(change) }
        if configuration.onSubmitCallback != 0 {
            events.append(.submit(
                callbackId: configuration.onSubmitCallback,
                nodeId: configuration.nodeID,
                value: draft
            ))
        }
        return events
    }

    mutating func clear() -> SearchFieldRendererEvent? {
        guard !configuration.disabled else { return nil }
        draft = ""
        return commitIfNeeded()
    }

    mutating func serverPublished(_ tree: NativeUITree) {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else { return }
        let published = SearchFieldRendererConfiguration(node: node)
        configuration = published
        if isFocused {
            pendingServerValue = published.value == lastCommitted ? nil : published.value
        } else {
            draft = published.value
            lastCommitted = published.value
            pendingServerValue = nil
        }
    }

    private mutating func commitIfNeeded() -> SearchFieldRendererEvent? {
        guard draft != lastCommitted else { return nil }
        lastCommitted = draft
        guard configuration.onChangeCallback != 0 else { return nil }
        return .change(
            callbackId: configuration.onChangeCallback,
            nodeId: configuration.nodeID,
            value: draft
        )
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id { return node }
        for child in node.children {
            if let found = findNode(id: id, in: child) { return found }
        }
        return nil
    }
}

struct SearchFieldRenderer: View {
    @ObservedObject private var bridge = NativeUIBridge.shared
    @State private var state: SearchFieldRendererState
    private let events: SearchFieldRendererEvents

    init(node: NativeUINode, events: SearchFieldRendererEvents = .native) {
        _state = State(initialValue: SearchFieldRendererState(node: node))
        self.events = events
    }

    var body: some View {
        FirstlightSearchFieldControl(
            configuration: state.configuration,
            text: Binding(
                get: { state.draft },
                set: { value in
                    if let event = state.userChanged(value) { events.send(event) }
                }
            ),
            onClear: {
                if let event = state.clear() { events.send(event) }
            },
            onSubmit: {
                for event in state.submit() { events.send(event) }
            },
            onFocusChanged: { focused in
                if let event = state.focusChanged(focused) { events.send(event) }
            }
        )
        .frame(minHeight: 44)
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
        .task(id: state.draft) {
            guard state.configuration.syncMode == "debounce" else { return }
            try? await Task.sleep(for: .milliseconds(state.configuration.debounceMilliseconds))
            if !Task.isCancelled, let event = state.flush() { events.send(event) }
        }
    }
}

File: resources/android/SearchFieldRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.TextFieldValue
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import kotlinx.coroutines.delay

sealed interface SearchFieldRendererEvent {
    val wireName: String

    data class Change(val callbackId: Int, val nodeId: Int, val value: String) : SearchFieldRendererEvent {
        override val wireName = "TEXT_CHANGE"
    }

    data class Submit(val callbackId: Int, val nodeId: Int, val value: String) : SearchFieldRendererEvent {
        override val wireName = "SUBMIT"
    }
}

class SearchFieldRendererEvents(private val send: (SearchFieldRendererEvent) -> Unit) {
    fun dispatch(event: SearchFieldRendererEvent) = send(event)

    companion object {
        val native = SearchFieldRendererEvents { event ->
            when (event) {
                is SearchFieldRendererEvent.Change -> NativeUIBridge.sendTextChangeEvent(event.callbackId, event.nodeId, event.value)
                is SearchFieldRendererEvent.Submit -> NativeUIBridge.sendSubmitEvent(event.callbackId, event.nodeId, event.value)
            }
        }
    }
}

data class SearchFieldRendererConfiguration(
    val nodeId: Int,
    val value: String,
    val placeholder: String,
    val disabled: Boolean,
    val autocapitalize: String,
    val autocorrectPolicy: String,
    val syncMode: String,
    val debounceMilliseconds: Int,
    val accessibilityLabel: String,
    val accessibilityHint: String,
    val onChangeCallback: Int,
    val onSubmitCallback: Int,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        value = node.props.getString("value"),
        placeholder = node.props.getString("placeholder"),
        disabled = node.props.getBool("disabled"),
        autocapitalize = node.props.getString("autocapitalize"),
        autocorrectPolicy = node.props.getString("autocorrect_policy", "default"),
        syncMode = node.props.getString("sync_mode", "live"),
        debounceMilliseconds = node.props.getInt("debounce_ms", 300).coerceAtLeast(50),
        accessibilityLabel = node.props.getString("a11y_label"),
        accessibilityHint = node.props.getString("a11y_hint"),
        onChangeCallback = node.props.getCallbackId("on_change"),
        onSubmitCallback = node.props.getCallbackId("on_submit"),
    )
}

class SearchFieldRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(SearchFieldRendererConfiguration(node)); private set
    var draft by mutableStateOf(TextFieldValue(configuration.value)); private set
    var lastCommitted = configuration.value; private set
    var pendingServerValue: String? = null; private set
    var focused = false; private set

    fun userChanged(value: String): SearchFieldRendererEvent? = userChanged(draft.copy(text = value))

    fun userChanged(value: TextFieldValue): SearchFieldRendererEvent? {
        if (configuration.disabled) return null
        draft = value
        return if (configuration.syncMode == "live") commit() else null
    }

    fun focusChanged(value: Boolean): SearchFieldRendererEvent? {
        focused = value
        return if (value) null else commit()
    }

    fun flush(): SearchFieldRendererEvent? = commit()

    fun submit(): List<SearchFieldRendererEvent> {
        if (configuration.disabled) return emptyList()
        return buildList {
            commit()?.let(::add)
            if (configuration.onSubmitCallback != 0) {
                add(SearchFieldRendererEvent.Submit(
                    configuration.onSubmitCallback,
                    configuration.nodeId,
                    draft.text,
                ))
            }
        }
    }

    fun clear(): SearchFieldRendererEvent? {
        if (configuration.disabled) return null
        draft = TextFieldValue("")
        return commit()
    }

    fun serverPublished(tree: NativeUITree) {
        val published = tree.root.findSearchFieldNode(configuration.nodeId)
            ?.let(::SearchFieldRendererConfiguration)
            ?: return
        configuration = published
        if (focused) {
            pendingServerValue = published.value.takeUnless { it == lastCommitted }
        } else {
            draft = TextFieldValue(published.value)
            lastCommitted = published.value
            pendingServerValue = null
        }
    }

    private fun commit(): SearchFieldRendererEvent? {
        if (draft.text == lastCommitted) return null
        lastCommitted = draft.text
        return configuration.onChangeCallback.takeIf { it != 0 }?.let {
            SearchFieldRendererEvent.Change(it, configuration.nodeId, draft.text)
        }
    }
}

object SearchFieldRenderer {
    @Composable
    fun Render(
        node: NativeUINode,
        modifier: Modifier,
        events: SearchFieldRendererEvents = SearchFieldRendererEvents.native,
    ) {
        val state = remember(node.id) { SearchFieldRendererState(node) }
        NativeUIBridge.currentTree.value?.let { tree ->
            LaunchedEffect(tree) { state.serverPublished(tree) }
        }
        LaunchedEffect(state.draft.text) {
            if (state.configuration.syncMode == "debounce") {
                delay(state.configuration.debounceMilliseconds.toLong())
                state.flush()?.let(events::dispatch)
            }
        }

        FirstlightSearchFieldControl(
            configuration = state.configuration,
            draft = state.draft,
            onValueChange = { state.userChanged(it)?.let(events::dispatch) },
            onClear = { state.clear()?.let(events::dispatch) },
            onSubmit = { state.submit().forEach(events::dispatch) },
            modifier = modifier.onFocusChanged {
                state.focusChanged(it.isFocused)?.let(events::dispatch)
            },
        )
    }
}

internal fun searchFieldKeyboardOptions(configuration: SearchFieldRendererConfiguration) = KeyboardOptions(
    capitalization = when (configuration.autocapitalize) {
        "none" -> KeyboardCapitalization.None
        "words" -> KeyboardCapitalization.Words
        "characters" -> KeyboardCapitalization.Characters
        else -> KeyboardCapitalization.Sentences
    },
    autoCorrectEnabled = when (configuration.autocorrectPolicy) {
        "enabled" -> true
        "disabled" -> false
        else -> null
    },
    imeAction = ImeAction.Search,
)

internal fun searchFieldKeyboardActions(submit: () -> Unit) = KeyboardActions(onSearch = { submit() })

private fun NativeUINode.findSearchFieldNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findSearchFieldNode(id)?.let { return it } }
    return null
}
